### ISSUE TITLE
"Unknown error while getting elements" error when option qt.process_model is process-per-site

### ISSUE THREAD

**Issue by cpixl** [qutebrowser/qutebrowser#5569]
**Version info**:
qutebrowser v1.12.0
Git commit: 
Backend: QtWebEngine (Chromium 80.0.3987.163)
Qt: 5.15.0 (compiled 5.13.2)

**Does the bug happen if you start with `--temp-basedir`?**:
Yes.

**Description**
When the option `qt.process_model` is set to `process-per-site`, the default `f` keybind doesn't always work.

**How to reproduce**
1. Run `qutebrowser  --base-tempdir --set qt.process_model process-per-site`
2. Type `:open example.com`, then press Enter
3. Type `:open !g test`, then press Enter
4. Press `f`, you should get an error message saying "Unknown error while getting elements" instead of the highlighted links

**Comment by toofar** [comment]
Nice work finding a reproducer. I use that setting and have seen it occasionally (on 5.14.2) when I have lots of github tabs open. I'm also seeing a `TypeError: Cannot read property 'webelem' of undefined`. I think that is from when HintManager calls `webelem.css_selector()` because there is no `window._qutebrowser`. It seems something to do with the redirect from ddg bang, I'm not even seeing where the early JS is injected from a normal navigation though...

**Comment by The-Compiler** [comment]
@toofar `window._qutebrowser` gets set from `qutebrowser/javascript/global_wrapper.js`, which is applied in `qutebrowser.utils.javascript.wrap_global()`, and that gets called in `qutebrowser.browser.webengine.webenginetab._WebEngineScripts.init` (which then uses `_inject_early_js`, which uses `QWebEngineScript`).

**Comment by The-Compiler** [comment]
And FWIW I think your analysis is correct, as shown by `:jseval --world application '_qutebrowser' in window` - that shows true initially, but false after that `!g test` call. Smells like a QtWebEngine bug to me though...

@toofar Are you also running with `qt.process_model process-per-site`?

**Comment by toofar** [comment]
> Are you also running with qt.process_model process-per-site?

Yep, an effort to reduce memory usage from back when I had less.

> and that gets called in qutebrowser.browser.webengine.webenginetab._WebEngineScripts.init

But that only gets called when a tab gets instantiated right? What about when you navigate somewhere? Don't we expect the `window` object to be a fresh instance after the navigation? I'm missing something because it clearly works most of the time.

**Comment by The-Compiler** [comment]
> But that only gets called when a tab gets instantiated right? What about when you navigate somewhere? Don't we expect the `window` object to be a fresh instance after the navigation? I'm missing something because it clearly works most of the time.

`QWebEngineScript` takes care of injecting the script for every visited page (hence why I believe this is a QtWebEngine bug). This is similar to how the Greasemonkey support works for Qt 5.9+, where Qt takes care of injecting the script based on the Greasemonkey comments, except it always gets injected if there is no such comment.

**Comment by lucilanga** [comment]
I can still see this very often with single-process and 6.9.0.

**Comment by rnhmjoj** [comment]
> I can still see this very often with single-process and 6.9.0.

Strange. Even in a clean profile? Try with running qutebrowser as:

qutebrowser -T -s qt.chromium.process_model single-process

For me the issue seems completely gone.

**Comment by lucilanga** [comment]
I confirm:

qutebrowser -T -s qt.chromium.process_model single-process and 6.9.0:

open a tab -> load an url -> press 'f' to see the hints

close tab -> undo close tab (:undo) -> press 'f' to see "Unknown error while getting elements"

**Comment by rnhmjoj** [comment]
Ok, I can reproduce that. However, the issue as I originally [reported](https://github.com/qutebrowser/qutebrowser/issues/6277), is gone.
